File: services/ai-agent/scripts/inspect_simple_collections.py
# Usage: python scripts/inspect_simple_collections.py
# Summarize fields in simple_* collections and check against expected webhook fields.

import json
from collections import defaultdict
import sys, pathlib, os, re

EXPECTED_FIELDS = {
    "_id",
    "full_name",
    "first_name",
    "last_name",
    "dob",
    "county",
    "booking_date",
    "bond",
    "bond_amount",
    "booking_number",
    "normalized_at",
    "category",
}

SIMPLE = ["harris", "brazoria", "galveston", "fortbend"]


def type_of(v):
    import bson
    if v is None:
        return "null"
    if isinstance(v, list):
        return "array"
    if isinstance(v, bson.ObjectId):
        return "ObjectId"
    return type(v).__name__


def summarize_collection(db, name: str):
    col = db[name]
    total = col.estimated_document_count()
    if total == 0:
        return {"name": name, "total": 0, "fields": {}, "samples": []}

    # Try to sort by booking_date if present, else sample
    try:
        cur = col.find({}).sort([("booking_date", -1)]).limit(20)
    except Exception:
        cur = col.aggregate([{"$sample": {"size": 20}}])

    fields = defaultdict(set)
    samples = []
    for i, doc in enumerate(cur):
        samples.append(doc)
        for k, v in doc.items():
            fields[k].add(type_of(v))

    field_summary = {k: sorted(list(v)) for k, v in fields.items()}

    # Alignment check
    present = set(field_summary.keys())
    missing = sorted(list(EXPECTED_FIELDS - present))
    extras = sorted(list(present - EXPECTED_FIELDS))

    return {
        "name": name,
        "total": total,
        "fields": field_summary,
        "present_expected": sorted(list(present & EXPECTED_FIELDS)),
        "missing_expected": missing,
        "extra_fields": extras,
        "sampleCount": len(samples),
        "samples_preview": [
            {k: doc.get(k) for k in [
                "_id","full_name","first_name","last_name","dob",
                "county","booking_date","bond","bond_amount",
                "booking_number","normalized_at","category"
            ] if k in doc}
            for doc in samples[:3]
        ]
    }


def read_env(path: pathlib.Path) -> dict:
    env = {}
    if not path.exists():
        return env
    line_re = re.compile(r"^\s*([A-Za-z_][A-Za-z0-9_]*)\s*=\s*(.*)\s*$")
    with path.open() as f:
        for raw in f:
            s = raw.strip()
            if not s or s.startswith('#'):
                continue
            m = line_re.match(s)
            if not m:
                continue
            k, v = m.group(1), m.group(2)
            # Strip surrounding quotes if present
            if (v.startswith('"') and v.endswith('"')) or (v.startswith("'") and v.endswith("'")):
                v = v[1:-1]
            env[k] = v
    return env


def main():
    # Load Mongo config from .env without instantiating app settings
    ROOT = pathlib.Path(__file__).resolve().parent.parent
    env = read_env(ROOT / '.env')
    mongo_uri = env.get('MONGO_URI') or os.environ.get('MONGO_URI')
    mongo_db = env.get('MONGO_DB') or os.environ.get('MONGO_DB')
    if not mongo_uri or not mongo_db:
        print(json.dumps({
            "error": "Missing MONGO_URI or MONGO_DB in .env or environment",
            "have_MONGO_URI": bool(mongo_uri),
            "have_MONGO_DB": bool(mongo_db)
        }, indent=2))
        sys.exit(2)

    from pymongo import MongoClient
    client = MongoClient(mongo_uri)
    _db = client[mongo_db]

    names = [f"simple_{c}" for c in SIMPLE if f"simple_{c}" in _db.list_collection_names()]

    report = {
        "database": _db.name,
        "collections": []
    }

    if not names:
        report["note"] = "No simple_* collections found"
    else:
        for n in names:
            report["collections"].append(summarize_collection(_db, n))

    print(json.dumps(report, default=str, indent=2))


if __name__ == "__main__":
    main()

File: services/ai-agent/scripts/inspect_simple_collections.mjs
// Usage: mongosh "$MONGO_URI/$MONGO_DB" inspect_simple_collections.mjs
// Summarizes fields and sample docs in simple_* collections used by Telnyx tools

const SIMPLE = ["harris", "brazoria", "galveston", "fortbend"];

function isObjectId(val) {
  return val && typeof val === 'object' && val._bsontype === 'ObjectId';
}

function typeOfVal(v) {
  if (v === null) return 'null';
  if (Array.isArray(v)) return 'array';
  if (isObjectId(v)) return 'ObjectId';
  return typeof v;
}

function summarizeCollection(db, name) {
  const col = db.collection(name);
  const total = col.estimatedDocumentCount();
  if (total === 0) {
    return { name, total, fields: {}, samples: [] };
  }

  // Sample a handful of recent docs if booking_date exists; else random
  let cursor;
  if (col.indexes().some(ix => ix.key && ix.key.booking_date)) {
    cursor = col.find({}).sort({ booking_date: -1 }).limit(10);
  } else {
    cursor = col.aggregate([{ $sample: { size: 10 } }]);
  }

  const fields = {};
  const samples = [];

  cursor.forEach(doc => {
    samples.push(doc);
    Object.keys(doc).forEach(k => {
      const t = typeOfVal(doc[k]);
      fields[k] = fields[k] || new Set();
      fields[k].add(t);
    });
  });

  // Convert sets to arrays and sort
  const fieldSummary = Object.fromEntries(
    Object.entries(fields).map(([k, set]) => [k, Array.from(set).sort()])
  );

  return { name, total, fields: fieldSummary, sampleCount: samples.length, samples };
}

function main() {
  const dbName = db.getName();
  print(`Inspecting simple_* collections in database: ${dbName}`);

  const names = db.getCollectionNames().filter(n => SIMPLE.map(c => `simple_${c}`).includes(n));
  if (names.length === 0) {
    print("No simple_* collections found.");
    return;
  }

  names.forEach(n => {
    const summary = summarizeCollection(db, n);
    print("\n==== "+n+" ====");
    printjson({ total: summary.total, fields: summary.fields, sampleCount: summary.sampleCount });
    // Print a few truncated samples (omit large arrays/strings)
    const showKeys = ["_id","full_name","first_name","last_name","dob","county","booking_date","bond","bond_amount","booking_number","normalized_at","category"]; 
    summary.samples.slice(0,3).forEach((s, i) => {
      const view = {};
      showKeys.forEach(k => { if (k in s) view[k] = s[k]; });
      print("-- sample "+(i+1)+" --");
      printjson(view);
    });
  });
}

main();
